src/modules/mc_pos_control/PositionControl/PositionControl.cpp

```diff
@@ -0,0 +1,241 @@
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file PositionControl.cpp
+ */
+
+#include "PositionControl.hpp"
+#include <float.h>
+#include <mathlib/mathlib.h>
+#include <ControlMath.hpp>
+#include <px4_defines.h>
+
+using namespace matrix;
+
+void PositionControl::setVelocityGains(const Vector3f &P, const Vector3f &I, const Vector3f &D)
+{
+	_gain_vel_p = P;
+	_gain_vel_i = I;
+	_gain_vel_d = D;
+}
+
+void PositionControl::setVelocityLimits(const float vel_horizontal, const float vel_up, const float vel_down)
+{
+	_lim_vel_horizontal = vel_horizontal;
+	_lim_vel_up = vel_up;
+	_lim_vel_down = vel_down;
+}
+
+void PositionControl::setThrustLimits(const float min, const float max)
+{
+	// make sure the thrust has a tiny minimal length to infer the direction
+	_lim_thr_min = math::max(min, -10e-4f);
+	_lim_thr_max = max;
+}
+
+void PositionControl::setState(const PositionControlStates &states)
+{
+	_pos = states.position;
+	_vel = states.velocity;
+	_yaw = states.yaw;
+	_vel_dot = states.acceleration;
+}
+
+void PositionControl::setInputSetpoint(const vehicle_local_position_setpoint_s &setpoint)
+{
+	_pos_sp = Vector3f(setpoint.x, setpoint.y, setpoint.z);
+	_vel_sp = Vector3f(setpoint.vx, setpoint.vy, setpoint.vz);
+	_acc_sp = Vector3f(setpoint.acceleration);
+	_thr_sp = Vector3f(setpoint.thrust);
+	_yaw_sp = setpoint.yaw;
+	_yawspeed_sp = setpoint.yawspeed;
+}
+
+void PositionControl::setConstraints(const vehicle_constraints_s &constraints)
+{
+	_constraints = constraints;
+
+	// For safety check if adjustable constraints are below global constraints. If they are not stricter than global
+	// constraints, then just use global constraints for the limits.
+	if (!PX4_ISFINITE(constraints.tilt) || (constraints.tilt > _lim_tilt)) {
+		_constraints.tilt = _lim_tilt;
+	}
+
+	if (!PX4_ISFINITE(constraints.speed_up) || (constraints.speed_up > _lim_vel_up)) {
+		_constraints.speed_up = _lim_vel_up;
+	}
+
+	if (!PX4_ISFINITE(constraints.speed_down) || (constraints.speed_down > _lim_vel_down)) {
+		_constraints.speed_down = _lim_vel_down;
+	}
+
+	if (!PX4_ISFINITE(constraints.speed_xy) || (constraints.speed_xy > _lim_vel_horizontal)) {
+		_constraints.speed_xy = _lim_vel_horizontal;
+	}
+}
+
+bool PositionControl::update(const float dt)
+{
+	_positionControl();
+	_velocityControl(dt);
+
+	_yawspeed_sp = PX4_ISFINITE(_yawspeed_sp) ? _yawspeed_sp : 0.f;
+	_yaw_sp = PX4_ISFINITE(_yaw_sp) ? _yaw_sp : _yaw;
+
+	const bool valid_acceleration_setpoint = PX4_ISFINITE(_acc_sp(0))
+			&& PX4_ISFINITE(_acc_sp(1))
+			&& PX4_ISFINITE(_acc_sp(2));
+	return valid_acceleration_setpoint;
```

Comment:
maybe rename to `controller_output_is_valid`. It took me a few seconds to understand why we checked the validity of the acceleration setpoints here. Also, the output is  `_thr_sp`; why don't we check the validity of this vector instead?

Reply:
The output getting executed is actually the attitude setpoint.

> It took me a few seconds to understand why we checked the validity of the acceleration setpoints here.

It's because any input setpoints and state combinations get processed but if at the end there was no valid acceleration vector there has to be at least one dimension in which the inputs were screwed up and the controller cannot generate valid output so this is an emergency.